### PR TITLE
Native sources build step output GraalVM version and docker image to use

### LINKED ISSUE #32328: native sources build step should tell the GraalVM version to use
### Description

When separating Java and native image compilation, I would like to know which GraalVM version is compatible.

Currently, we need to manually follow the version as described in the documentation and when using multiple version of Quarkus compatible with different versions of GraalVM it is not easy to have a single CI pipeline setup. As sometimes, newer versions of GraalVM are not compatible with older version of Quarkus, having one GraalVM version is not always an option.

Thus, it should be possible to get the GraalVM version that is recommended so we can install the correct one.
Ideally, when combining `-Dnative-sources` and `-Dquarkus.native.container-build=true`, it should output the docker image name that Quarkus plugin would have use. The idea is to be able to do : 

```sh
cd target/native-image
docker run \
  -it \
  --rm \
  --v $(pwd):/work 
  -w /work 
  --entrypoint bin/sh \
  $(cat native-builder.image) \ 
  -c "native-image $(cat native-image.args) -J-Xmx4g"
```

### Implementation ideas

`NativeImageBuildStep.nativeSourcesResult` could be augmented with something like

```
if(nativeConfig.isContainerBuild()){
            try (FileOutputStream dockerImageNameFOS = new FileOutputStream(outputDir.resolve("native-builder.image").toFile())) {
                dockerImageNameFOS.write(nativeConfig.getEffectiveBuilderImage().getBytes(StandardCharsets.UTF_8));
            } catch (Exception e) {
                throw new RuntimeException("Failed to build native image sources", e);
            }
        }
        else {
            try (FileOutputStream dockerImageNameFOS = new FileOutputStream(outputDir.resolve("graalvm.version").toFile())) {
                dockerImageNameFOS.write(GraalVM.Version.CURRENT.version.toString().getBytes(StandardCharsets.UTF_8));
            } catch (Exception e) {
                throw new RuntimeException("Failed to build native image sources", e);
            }
        }
```

### PR DESCRIPTION
**Pull request by scrocquesel**
Resolves: #32328

### CODE REVIEW

File: core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java [hunk 1]
```diff
@@ -142,16 +142,28 @@ ArtifactResultBuildItem nativeSourcesResult(NativeConfig nativeConfig,
...
         }
 
+        if (nativeConfig.isContainerBuild()) {
+            try (FileOutputStream dockerImageNameFOS = new FileOutputStream(
+                    outputDir.resolve("native-builder.image").toFile())) {
+                dockerImageNameFOS.write(nativeConfig.getEffectiveBuilderImage().getBytes(StandardCharsets.UTF_8));
+            } catch (Exception e) {
+                throw new RuntimeException("Failed to build native image sources", e);
+            }
+        }
```
Comment (gsmet) [inline core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java:162]: Maybe put it in the same try above. Or is there a reason to keep it separate?

File: core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java [hunk 2]
```diff
@@ -142,16 +142,28 @@ ArtifactResultBuildItem nativeSourcesResult(NativeConfig nativeConfig,
...
-                    + outputDir);
+            dockerImageNameFOS.write(GraalVM.Version.CURRENT.version.toString().getBytes(StandardCharsets.UTF_8));
         } catch (Exception e) {
             throw new RuntimeException("Failed to build native image sources", e);
         }
 
+        if (nativeConfig.isContainerBuild()) {
+            try (FileOutputStream dockerImageNameFOS = new FileOutputStream(
+                    outputDir.resolve("native-builder.image").toFile())) {
+                dockerImageNameFOS.write(nativeConfig.getEffectiveBuilderImage().getBytes(StandardCharsets.UTF_8));
```
Comment (gsmet) [inline core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java:158]: Maybe use `Files.write()` instead? And we can do the same above?

A simple try/catch with all 3 Files.write() should be the best outcome?

You can do it in a separate commit given it's a refactoring of an existing code.

**Review by gsmet — COMMENTED** [review]
It looks like a good idea. I suggested a few improvements.

File: core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java (hunk 1 above)
Comment (scrocquesel) [reply core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java:162]: No reason apart keeping the try resource syntax

**Comment by gsmet** [comment]
Changes look good from my side. I will let @zakkak review it more carefully and merge if he's happy with it.

**Review by geoand — COMMENTED** [review]
I like the idea!

**Review by zakkak — APPROVED** [review]
LGTM and is a nice addition. Thanks!

FYI upstream GraalVM will introduce Native-Image build bundles (https://github.com/oracle/graal/issues/5473) which could potentially be used for the `native-sources` package type.

CI Failures seem unrelated.

**Pull request merged**